### DISCUSSION TITLE
[Question] Learning NixOS

### DISCUSSION THREAD

**Discussion by chris-kruining** [Icy-Thought/snowflake#26] [Q&A]
Hi,

I am improving my nixos setup by taking a close look at your repo. I've copied a lot of code already, but now my build is failing with `error: attribute 'mapModules' already defined` on rebuild, this error is triggered on `inherit (lib.my) mapModules mapModules mapHosts;` in my `flake.nix`

I do not get why this error is triggering though. I mean nix itself probably already has the function, but it is loaded from `lib.my` right?

Do you have any idea what might cause this?

Edit: took me a bit to get pushing to my repo up and running again, so for context; this is what I have thusfar: https://github.com/chris-kruining/.files

**Comment by Icy-Thought** [comment]
Don't know if you noticed this, but you have two entries of `mapModules`. The second one should be `mapModulesRec`

**Reply by chris-kruining** [reply]
I as a matter of fact have not seen that, at all, despite me staring at that particular line for ages....

thanks!

**Reply by Icy-Thought** [reply]
It happens to the best of us! I encountered a similar issue yesterday too, where I'd accidentally forgotten to place `PATH` for cargo inside `sessionPath` instead of `sessionVariables` and it had me confused why nothing would work only to notice that I had wiped the path completely when setting it for cargo and later built on top of it. :facepalm:  

Nonetheless, glad to help out! :)

**Comment by chris-kruining** [comment]
alright, next question. with the config we have for all of our modules, how can I apply my config to multiple users?

for example, with my old setup I had also applied zsh with starship for root. but that is missing now, nor is git enabled for root.

**Reply by Icy-Thought** [reply]
Never worked with multiple users and Nix, so I honestly don't know. But you made me curious to try it out and perhaps even use it myself for a separate environment for work + fidgeting around.

I will research this bit and perhaps even come up with a solution -> share it with you! But don't expect anything soon since I have been very busy recently, and because of this I cannot guarantee when I'll be able to fidget with this idea.

**Reply by chris-kruining** [reply]
not looking for multi user per se, more wanting to also apply all my terminal stuff for root as well, so that whenever I do a `sudo su` I don't get dropped of a cliff.

**Reply by Icy-Thought** [reply]
Out of curiosity, why would you want to `sudo su`?

**Reply by Icy-Thought** [reply]
Also, you could always replace the home-manager bits with nix once since they do work system-wide. (iirc)

**Reply by chris-kruining** [reply]
> Out of curiosity, why would you want to `sudo su`?

so that I change to the root user, so that I dont have to prefix all of my commands with sudo and fill in the password. and no, I don't need root for everything, but a `nixos-rebuild` needs sudo.



> Also, you could always replace the home-manager bits with nix once since they do work system-wide. (iirc)

not all the config is compatible both ways right? I tried that with my old setup and ran into a lot of issues. But then again, I am a noob at nix so perhaps I was doing something wrong

**Reply by Icy-Thought** [reply]
> but a nixos-rebuild needs sudo.

That's why I have my `sysdate` command, which translates to `nixos-rebuild switch --use-remote-sudo --flake .#"$(hostname)" --impure`.
And you can always just use `sudo !!~ (does not work in non posix compliant shells) to apply sudo to the previous command!

> not all the config is compatible both ways right?

I honestly don't know, since I never spent any time in either `sudo su` or `sudo -i`. But I can come back on that one a little later today and see if I can get it to work.
